docs/jax-101/01-jax-basics.md

```diff
@@ -0,0 +1,377 @@
+
+xs = np.random.normal(size=(100,))
+noise = np.random.normal(scale=0.1, size=(100,))
+ys = xs * 3 - 1 + noise
+
+plt.scatter(xs, ys)
+```
+
++++ {"id": "RTh22mo4rR1x"}
+
+Therefore, our model is $\hat y(x; \theta) = wx + b$.
+
+We will use a single array, `theta = [w, b]` to house both parameters:
+
+```{code-cell} ipython3
+:id: TnVrRTMamyzb
+
+def model(theta, x):
+  """Computes wx + b on a batch of input x."""
+  w, b = theta
+  return w * x + b
+```
+
++++ {"id": "qCrLmmKrn9_h"}
+
+The loss function is $J(x, y; \theta) = (\hat y - y)^2$.
+
+```{code-cell} ipython3
+:id: 07eMcDLMn9Ww
+
+def loss_fn(theta, x, y):
+  prediction = model(theta, x)
+  return jnp.mean((prediction-y)**2)
+```
+
++++ {"id": "ejMt4dulnoYX"}
+
+How do we optimize a loss function? Using gradient descent. At each update step, we will find the gradient of the loss w.r.t. the parameters, and take a small step in the direction of steepest descent:
+
+$\theta_{new} = \theta - 0.1 (\nabla_\theta J) (x, y; \theta)$
+
+```{code-cell} ipython3
+:id: 2I6T5Wphpaaa
+
+def update(theta, x, y, lr=0.1):
+  return theta - lr * jax.grad(loss_fn)(theta, x, y)
+```
+
++++ {"id": "MAUL1gT_opVn"}
+
+In JAX, it's common to define an `update()` function that is called every step, taking the current parameters as input and returning the new parameters. This is a natural consequence of JAX's functional nature, and is explained in more detail in [The Problem of State](https://colab.research.google.com/github/google/jax/blob/master/docs/jax-101/07-state.ipynb).
+
+This function can then be JIT-compiled in its entirety for maximum efficiency. The next guide will explain exactly how `jax.jit` works, but if you want to, you can try adding `@jax.jit` before the `update()` definition, and see how the training loop below runs much faster.
+
+```{code-cell} ipython3
+:id: WLZxY7nIpuVW
+
+theta = jnp.array([1., 1.])
+
+for _ in range(1000):
+  theta = update(theta, xs, ys)
+
+plt.scatter(xs, ys)
+plt.plot(xs, model(theta, xs))
+
+w, b = theta
+print(f"w: {w:<.2f}, b: {b:<.2f}")
+```
+
++++ {"id": "5-q17kJ_rjLc"}
+
+As you will see going through these guides, this basic recipe underlies almost all training loops you'll see implemented in JAX. The main difference between this example and real training loops is the simplicity of our model: that allows us to use a single array to house all our parameters. We cover managing more parameters in the later [pytree guide](https://colab.research.google.com/github/google/jax/blob/master/docs/jax-101/06-pytrees.ipynb). Feel free to skip forward to that guide now to see how to manually define and train a simple MLP in JAX.
+
+Another important difference is that our simple model could be defined manually. To help define more complicated models like neural nets, you'll find it useful to learn a higher-level library like Haiku, which we also cover in a [later guide](https://colab.research.google.com/github/google/jax/blob/master/docs/jax-101/08-haiku-basics.ipynb).
```

Comment:
I must have missed this in my first review, but let's not mention Haiku or a later NN guide for now, since we're not sure what the story is there yet. I think it's fine to just delete this paragraph.

Reply:
Done, thanks!